overlord/devicestate/devicemgr.go

```diff
@@ -1802,50 +1802,77 @@ func (m *DeviceManager) Systems() ([]*System, error) {
 	return systems, nil
 }
 
-// SystemAndGadgetInfo return the system details including the model
-// assertion and gadget details for the given system label.
-func (m *DeviceManager) SystemAndGadgetInfo(wantedSystemLabel string) (*System, *gadget.Info, error) {
+// SystemAndGadgetAndEncryptionInfo return the system details
+// including the model assertion, gadget details and encryption info
+// for the given system label.
+func (m *DeviceManager) SystemAndGadgetAndEncryptionInfo(wantedSystemLabel string) (*System, *gadget.Info, *EncryptionSupportInfo, error) {
 	if m.isClassicBoot {
-		return nil, nil, fmt.Errorf("cannot get model and gadget information on a classic boot system")
+		return nil, nil, nil, fmt.Errorf("cannot get model and gadget information on a classic boot system")
 	}
 	// get current system as input for loadSeedAndSystem()
 	systemMode := m.SystemMode(SysAny)
 	currentSys, _ := currentSystemForMode(m.state, systemMode)
 
 	s, sys, err := loadSeedAndSystem(wantedSystemLabel, currentSys)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// 2. get the gadget volumes for the given seed-label
 	perf := &timings.Timings{}
-	if err := s.LoadEssentialMeta([]snap.Type{snap.TypeGadget}, perf); err != nil {
-		return nil, nil, fmt.Errorf("cannot load gadget snap metadata: %v", err)
+	if err := s.LoadEssentialMeta([]snap.Type{snap.TypeKernel, snap.TypeGadget}, perf); err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot load gadget snap metadata: %v", err)
 	}
-	gadgetSnap := s.EssentialSnaps()[0]
-	if gadgetSnap.Path == "" {
-		return nil, nil, fmt.Errorf("internal error: cannot get gadget snap path")
+	// EssentialSnaps is always ordered, see asserts.Model.EssentialSnaps:
+	// "snapd, kernel, boot base, gadget."
+	//
+	// kernel info needed to check encryptionSupport
+	kernelSnap := s.EssentialSnaps()[0]
```

Comment:
I guess I am missing something here, why do we know that there is no snapd snap so 0 index is the kernel? Same question for gadget and index 1, which assumes there is no base.

Reply:
I updated the comment a bit, hopefully it's more clear now? Essentially the order is always the same but if some snap types are not loaded in LoadEssentialMeta() a few lines above they are skipped and the indexes changes but the order remains.

Reply:
That clarifies things indeed, thank you